Prepullrequest

tymochko opened the pull request:
Our latest final changes

mnkami commented:
Wow :) hard to check everything in it

mnkami commented on .gitignore:
```diff
@@ -41,4 +41,6 @@ jspm_packages
 
 dist
 public/
+src/server/usersInfo/
```
Why usersInfo?! If we keep fles there it shoukd be moved to sercer/static/users/

mnkami commented on gulpfile.js:
```diff
@@ -5,20 +5,26 @@ const babel = require('gulp-babel');
-        }))
-        .pipe(gulp.dest('public/scripts'));
+    // gulp.src('src/server/app.js')
+    //     .pipe(babel({
```
Commented code

mnkami commented on src/client/index.html:
```diff
@@ -9,30 +9,15 @@
     <!-- endinject -->
 
-    <!-- <link href="public/stylesheets/style.css" rel="stylesheet"> -->
-    <!-- <link href="node_modules/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet"> -->
```
Nice

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+        })();
+        
+    }
+// });
```
This file needs cleanup an refactoring, but you know that

mnkami commented on src/client/modules/profile/avatarContent.html:
```diff
@@ -9,18 +9,20 @@ <h3 class="modal-title">{{currentImage.description}}</h3>
     <div class="modal-footer">
         <button class="btn btn-primary" type="button" ng-click="ok()">Yes</button>
         <button class="btn btn-warning" type="button" ng-click="cancel()">No</button>
+        <button class="btn btn-success pull-left" type="button" ng-click="uploadPhoto()">Upload photo</button>
```
(Optional) Upload avatar sounds better

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,70 +1,98 @@
-    };
+    .service('profileService', ['$http', function ($http) {
+        let userId = '';
+        const profileURL = 'api/users/profile';
```
Would b enice to move out every url like this to a separate file with constants

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,70 +1,98 @@
+                url: 'api/users/profile/upload',
+                data: {file: file}
+            }).then((resp) => {
+                uploadedImg = resp.data;
```
Global variable! Rid of it.

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,70 +1,98 @@
+                controller: 'uploadController'
+            });
+            uploadInstance.result.then((img) => {
+                $scope.customImages = [];
```
$scope.customImages = [img];

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+                angle = tiltTank(xCoordinate);
+                ctx.save();
+                if (angle > 0) {
+                    var angle_curr = angle / Math.PI * 180;
```
copy paste. use helper function `getAngleInRad(angle)`

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+                else {
+                    var angle_curr = -angle / Math.PI * 180;
+                    angle_curr = Math.round(angle_curr / 10);
+                    var xCof = angle_curr * 5.5;
```
5.5?! magic strings, rid of it, use named constants

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+                    yCof = yCof + (angleWeapon_curr * 6.5);
+                    
+                } 
+                else if (angle > angleWeapon){
```
what if angle is the same?

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+                    ctx.translate(xCoordinate + 48 - xCof, yCoordinate - 43 + yCof);
+                }    
+
+                else {
```
`ctx.translate(xCoordinate + 48 - xCof, angle > 0 ? yCoordinate - 43 + yCof : yCoordinate - 43 - yConf);`

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+
+            this.angle = Math.atan(tan);
+
+            // if (tan < 0)
```
commented code

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+        };
+
+        var clear = function() {
+            ctx.clearRect(0, 0, WIDTH, HEIGHT);
```
please use more specific names for the constants

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+            let a = (point2.y - point1.y) / (point2.x - point1.x);
+            let b = point1.y - a * point1.x;
+
+            if(Math.abs(objPoint.y - (a*objPoint.x + b)) < 2) {
```
just `return Math.abs(objPoint.y - (a*objPoint.x + b)) < 2`

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,869 @@
+
+            if ( (y - 5) <= segmentY && segmentY <= (y + 5) ) {
+                return y;
+            }
```
add `return null;`

tymochko replied:
I was thinking about it but in my opinion - there is not way that **y** could not be found at all.
Or it does not matter and should be returned always in all cases?

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,157 @@
+    .left-col {
+        display: inline-block;
+    }
+    // .right-col {
```
commented code

mnkami commented on src/server/api/users/usersRoutes.js:
```diff
@@ -112,4 +134,82 @@ router.put('/profile/delete', (req, res) => {
+            cb(null, dir);
+        },
+        filename: function (req, file, cb) {
+            cb(null, fileNameNew + '.' + file.originalname.split('.')[file.originalname.split('.').length - 1]);
```
this is is unacceptable and unreadable.

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -232,11 +242,116 @@ const deleteUser = function (id, callback) {
 
+            userImage = foundUser.userImg;
+            if(userImage.uploadedImg) {
+                userDir = __dirname + '/../../usersInfo/' + userId + '/' + userImage.image;
```
I have already said about this... use helper function

tymochko commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,867 @@
+                if (checkCol(coords,originalPoints)){
+                    console.log( 'x:' +  (coords.x + coords.width), 'y:' + (coords.y + coords.height));
+                    bullets.splice(i, 1);
+                    tick(coords);                                   // <------ Explosion ------>
```
May I suggest a better name for this function? Like `generateExplosion` for instance

mnkami commented on .gitignore:
```diff
@@ -41,4 +41,6 @@ jspm_packages
 
 dist
 public/
+src/server/static/usersInfo/
```
what other folders does static includes? afaics static does not present in git.

mnkami commented on gulpfile.js:
```diff
@@ -39,19 +53,34 @@ gulp.task('template', () => {
+
+gulp.task('test', function (done) {
+    new Server({
+        configFile:__dirname + '/karma.conf.js',
```
missing space after semicolon.

mnkami commented on gulpfile.js:
```diff
@@ -39,19 +53,34 @@ gulp.task('template', () => {
+    }, done).start();
+});
+
+gulp.task('js-models', () => {
```
if you are using es6 syntax - use it everywhere

mnkami commented on gulpfile.js:
```diff
@@ -39,19 +53,34 @@ gulp.task('template', () => {
+        .pipe(sourcemaps.init())
+        .pipe(concat('main-models.js'))
+        .pipe(sourcemaps.write())
         .pipe(gulp.dest('public/'));
```
create a constant for "public/" -- it used here more than once.

mnkami commented on karma.conf.js:
```diff
@@ -0,0 +1,97 @@
+        // test results reporter to use
+        // possible values: 'dots', 'progress'
+        // available reporters: https://npmjs.org/browse/keyword/karma-reporter
+        reporters: ['progress', 'html'],
```
https://github.com/karma-runner/karma-coverage
would be nice

mnkami commented on karma.conf.js:
```diff
@@ -0,0 +1,97 @@
+
+        // start these browsers
+        // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
+        browsers: ['Chrome'],
```
Browser for console is still missing, WTF?

mnkami commented on karma.conf.js:
```diff
@@ -0,0 +1,97 @@
+// Karma configuration
+// Generated on Mon Jul 11 2016 17:05:15 GMT+0300 (FLE Daylight Time)
```
reportSlowerThan: 100, -- one more useful config.

mnkami commented on src/client/models/generateDamage.js:
```diff
@@ -0,0 +1,317 @@
+        }
+    });
+
+    for (let i = 1; i < pointsOfIntersect.length; i++) {
```
Array.from({ length: pointsOfIntersect.length }, (val, key) => key); // [0, 1, 2, 3]
Just so you know...

tymochko replied:
Wow, this will take some time to process...

mnkami replied:
not a must. in case you have time and in current situation it does not make a big difference.

mnkami commented on src/client/models/generateDamage.js:
```diff
@@ -0,0 +1,317 @@
+
+    if (deltaX != 0) {
+        return ( (damagePointX - endpoint1X) / deltaX );
+    } else {
```
else is redundant, just return the value

tymochko replied:
Thanks.
Actually this is a mistake in code logic - already fixing it on my branch

mnkami commented on src/client/models/generateDamage.js:
```diff
@@ -0,0 +1,317 @@
+    deltaX = (endpoint2X - endpoint1X);
+    deltaY = (endpoint2Y - endpoint1Y);
+
+    if (deltaX != 0) {
```
!==

tymochko replied:
Thanks

mnkami commented on src/client/models/generateDamage.js:
```diff
@@ -0,0 +1,317 @@
+
+const findLineSegmentCoefficient = (endpoint1X, endpoint1Y, endpoint2X, endpoint2Y, damagePointX, damagePointY) => {
+    // find coefficient of point, situated on line segment
+    let deltaX,
```
I'd use `const deltaX = (endpoint2X - endpoint1X);`

tymochko replied:
I see.

mnkami commented on src/client/models/generateDamage.js:
```diff
@@ -0,0 +1,317 @@
+     * cX, cY and r - are coordinates of center of damage and a radius */
+
+    /* using line equation (y = m*x + k) */
+    let m = ( (y2 - y1) / (x2 - x1) );
```
review your let/const... a lot of `let` variables could be `const`.

tymochko replied:
Ok

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+    var lastFire = Date.now();
+    var gameTime = 0;
+    const g = 9.81;
+    var bulletImg=new Image();
```
no vars!

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+            angle: angle,
+            bulletSpeed: power
+        });
+        lastFire = Date.now();
```
it's not reliable to use clients time. In future please replace by server time.

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+        for(var i = 0; i < bullets.length; i++) {
+            bullet = bullets[i];
+            bullet.pos[0] = tankX + weaponWidth * Math.cos(angleWeapon + angle*Math.PI/180) + bullet.bulletSpeed * dt2*Math.cos(bullet.angle*Math.PI/180 + angleWeapon);
+            bullet.pos[1] = tankY-30 - weaponWidth * Math.sin(angleWeapon + angle*Math.PI/180)- (bullet.bulletSpeed * dt2*Math.sin(bullet.angle*Math.PI/180 + angleWeapon) - g * dt2 * dt2 / 2);
```
Try to limit your strings up to 80 symbols on a line.

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+            var groundPath = new paper.Path(
+                new paper.Point(originalPoints[0][0], originalPoints[0][1])
+                );
+            for(let i = 1; i < originalPoints.length; i++) {
```
do not use the same variable name in the inner loop

mnkami replied:
why not forEach?

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+                    y: intersect[0]._point.y
+                };
+                console.log( 'x:' +  crossPoint.x, 'y:' + crossPoint.y );
+                tick(crossPoint);                                   // <------ Explosion ------>
```
tick?! === explosion?!

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+            {
+                bullets.splice(i, 1);
+                window.cancelAnimationFrame(requestAnimFrame);
+                i--;
```
you are decremented a loop variable, not good...

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    function render() {
+        renderEntities(bullets);
```
please encapsulate the scope so you could test it

mnkami commented on src/client/models/shotTrajectory.js:
```diff
@@ -0,0 +1,161 @@
+    };
+
+    function renderEntities(list) {
+        for(var i=0; i<list.length; i++) {
```
`(list || []).forEach((item) => renderEntity)` so simple
or move fallback as default arg `renderEntities(list = []) {`

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -0,0 +1,53 @@
+                arr.push([Math.round(a), Math.round(b)]);
+            }
+            for(let i = 0; i < arr.length; i++) {
+                if(arr[i][0] === posX) return (arr[i][1]);
```
what if there is no such points? findLinePoints  should return smth.

mnkami replied:
always use braces

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -0,0 +1,53 @@
+        draw(timePassed);
+        if(tankX >= WIDTH - 11 || tankX <= 11){
+            window.cancelAnimationFrame(requestAnimFrame);
+            // console.log('stop!!!');
```
please clean up

kachanovskyi replied:
comment? or code itself is too messed up?

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -0,0 +1,53 @@
+        let timePassed = time - start;
+        if (timePassed > duration) timePassed = duration;
+        draw(timePassed);
+        if(tankX >= WIDTH - 11 || tankX <= 11){
```
magic number, I want it to be 100500

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -0,0 +1,53 @@
+        } else {
+            tankX--;
+        }
+        angle = parseInt(getId('angle').innerHTML);
```
why this angle is still not encapsulated?

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,304 @@
+                window.oRequestAnimationFrame      ||
+                window.msRequestAnimationFrame     ||
+                function(callback){
+                    window.setTimeout(callback, 1000 / 60);
```
please use const for time to wait

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,304 @@
+            colors.forEach(function (color) {
+                polygon(points, color);
+ 
+                // points = points.map(function(pair) {
```
commented code

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,304 @@
+ 
+        var drawSky = function(){
+            var grd=backCtx.createLinearGradient(0,0,0,500);
+            grd.addColorStop(0,"#172059");
```
please use consts for the colors

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,304 @@
+ 
+            this.angle = Math.atan(tan);
+ 
+            // if (tan < 0)
```
dead code

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -0,0 +1,304 @@
+            }
+            var tan = (y1 > y2) ? (y1 - y2) / (x1 - x2) : (y2 - y1) / (x2 - x1);
+ 
+            this.angle = Math.atan(tan);
```
why this?

mnkami commented on src/client/modules/chat/index.js:
```diff
@@ -1,77 +1,60 @@
+          if(data.length)
+          {
+               for(var x=data.length-1;x>=0; --x){
+                    $scope.$apply(function () {
```
as we discussed apply should be triggered only once, when all arrays is constructed

mnkami commented on src/client/modules/game/index.js:
```diff
@@ -0,0 +1,59 @@
+    initGame();
+}
+
+// function ChatController($scope,socket, $sce) {
```
Still not working?!

mnkami commented on src/client/modules/home/index.js:
```diff
@@ -1,103 +1,37 @@
+function MainCtrl($scope, $uibModal, $log, $location, $http, $window) {
+
+    //<------------slider------------->
+    $scope.myInterval = 2000;
```
where are u using it? seems like it's redundant...

kachanovskyi replied:
What exactly do you mean?

mnkami replied:
I haven't found usage for this variable (used github search)

kachanovskyi replied:
Still don't get it. Sorry
What variable?

shmoksi replied:
$scope.myInterval

kachanovskyi replied:
![untitled](https://cloud.githubusercontent.com/assets/19438192/16931590/5ee97630-4d4a-11e6-8499-3186ac06df57.png)

mnkami commented on src/client/modules/login/index.js:
```diff
@@ -1,73 +1,55 @@
+                    .then((response) => {
+                        loginResult.result = true;
+                        $uibModalInstance.close($scope.selected.item);
+                        $window.location.reload();
```
what for?

mnkami commented on src/client/modules/login/index.js:
```diff
@@ -1,73 +1,55 @@
+            };
+
+            sendLog.log(userInfo, $scope, $uibModalInstance, items, loginResult);
+            // alert(loginResult.result);
```
commented code

mnkami commented on src/client/modules/login/login.html:
```diff
@@ -6,7 +6,7 @@ <h2 class="form-header">Login</h2>
                     <div>
                     <label for="inputEmail">User name:</label>
-                        <input name="username" type="text" class="form-control" id="inputEmail" placeholder="Username" ng-model="user.name" ng-required="true" ng-minlength=5 ng-maxlength=15 ng-pattern="/^[a-zA-Z0-9]*$/">
+                        <input name="username" type="text" class="form-control" id="inputEmail" placeholder="Username" ng-model="user.name" ng-required="true" ng-minlength="minLengthName" ng-maxlength="maxLengthName" ng-pattern="/^[a-zA-Z0-9]*$/">
```
so is it a user name or an email?

mnkami commented on src/client/modules/navigation/index.js:
```diff
@@ -0,0 +1,156 @@
+                $http.get("/api/users/checkSession").then((res) => {
+                    if (res.data.status === 'success') {
+                        $scope.logged = true;
+                    } else if (res.data.status == 'error'){
```
just put else.

mnkami commented on src/client/modules/navigation/index.js:
```diff
@@ -0,0 +1,156 @@
+                        $scope.logged = false;
+                    }
+                }, () => {
+                    console.log('server error');
```
use should be able to see the error, not it console.

kachanovskyi replied:
alert will work? At least for now?

mnkami commented on src/client/modules/navigation/index.js:
```diff
@@ -0,0 +1,156 @@
+                navRight = '<ul class="nav navbar-nav navbar-right"> ';
+            if(logged) {
+                for(let i = 0; i < routeNavigation.routes.length; i++) {
+                    href = ' href="' + routeNavigation.routes[i].template + '"';
```
use es6 template literals, code will be cleaner 
https://developers.google.com/web/updates/2015/01/ES6-Template-Strings

kachanovskyi replied:
will do

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,68 +1,110 @@
+            }).then((resp) => {
+                uploadedImg = resp.data; // TODO delete 'uploadImg' and send directly resp.data
+                $uibModalInstance.close(uploadedImg);
+            });
```
`catch`?

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,68 +1,110 @@
+        $scope.mass = false;
+        
+        $scope.set = function (a) {
+            $scope.mass = a
```
; is missing

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,68 +1,110 @@
+    }])
+    .controller('deleteUserController', ['$scope', '$uibModalInstance', ($scope, $uibModalInstance) => {
+
+        $scope.test1 = false;
```
remove it

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,68 +1,110 @@
+            $scope.images = res.data;
+            $scope.setCurrentImage($scope.images[0]);
+        }).catch((err) => {
+            console.log(err);
```
should be proper error handling

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -87,33 +130,37 @@ function manageProfileController($scope, $uibModal, profileService, toastr, $loc
+
+
+    function getSalt() {
+        return "?salt=" + new Date().getTime();
```
what for? ignore cache?

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -122,64 +169,49 @@ function manageProfileController($scope, $uibModal, profileService, toastr, $loc
-            savingMsg();
-
 
+            userInfo.userOldPassword = user.oldPassword;
```
`Object.assign` would be perfect here

mnkami commented on src/client/modules/signup/index.js:
```diff
@@ -12,22 +12,12 @@ module.exports = angular.module('tanks.signup', [
-		return $http.post('http://localhost:3000/api/users/add', userInfo).then(function(res){
+		return $http.post('/api/users/add', userInfo).then(function(res){
 			$location.path('/');
 		});
```
error handling is missing

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
+     .right-col {
+         display: inline-block;
+         float: right;
+         background-color: #BA2924;
```
const

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
+        width: 100%;
+        padding-left: 20px;
+        padding-right: 20px;
+        // font-weight: bold;
```
commented ocde

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
-    background-color: #BA2924;
+    position: relative;
+    top: 410px; 
+    user-select: none;
```
mixin

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
+        height: 38px;
+        font-size: 35px;
+        font-family: sans-serif;
+        color: #09090F;
```
should be a constant

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
+        border: 2px dotted aqua;
+    }
+
+    .power-text, .fire-text, .angle-text {
```
optional: each selector from new line

mnkami commented on src/client/scss/_game.scss:
```diff
@@ -1,16 +1,176 @@
+        width: 400px;
+        height: 170px;
+        top: -30px;
+        left: 0px;
```
left: 0;

mnkami commented on src/client/scss/toastr.scss:
```diff
@@ -0,0 +1,273 @@
+  border-radius: $radius;
+}
+
+@mixin boxShadow($boxShadow){
```
bootstrap already has it

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -1,7 +1,21 @@
+const userInfoDir = './src/server/static/usersInfo/';
+const publicScopeName = 'public';
+const userUploadsScopeName = 'userUploads';
+const publicImgURL = "/api/users/profile/getImage/" + publicScopeName + '/';
```
`/api/users/profile/getImage/${publicScopeName}/`;

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -137,13 +153,26 @@ const logoutUser = function (id, callback) {
+const checkUser = function (id, callback) {
+    console.log('checkUser', id);
+    this.findOne(id, (err, foundUser) => {
+        // if (err) {
```
commented code

mnkami replied:
this.findOne(id, callback);

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -171,16 +200,16 @@ const updateUser = function (id, updatedData, callback) {
-                
-                    function(err, foundUser) {
+
+                    function (err, foundUser) {
```
just `callback`

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+    function cb(err, foundUser) {
+        userImage = foundUser.userImg;
+        if (userImage.uploadedImg) {
+            userDir = __dirname + '/../../static/usersInfo/' + userId + '/' + userImage.image;
```
3rd time I'm poiniting on this!!! use helper function

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+    var text = `Hello ${userName}! Welcome to PocketTanks game!`;
+
+    var mailOptions = {
+        from: 'pockettanksmail@gmail.com',
```
it should be situated in a config file.

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+        if (error) {
+            console.log(error);
+            ;
+        } else {
```
really?

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+        }
+    });
+    var upload = multer({ //multer settings
+        storage: storage
```
const upload = multer({ storage }).single("file")'

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+    upload(request, res, function (err) {
+        if (err) {
+
+            ress.json({error_code: 1, err_desc: err});
```
typo, no `ress` defined

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+};
+
+const getPublicImg = function (req, res) {
+    fs.readdir(__dirname + '/../../static/images/' + '/', function (e, files) {
```
`+ '/'` is uselsess

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -233,10 +262,164 @@ const deleteUser = function (id, callback) {
+    });
+};
+
+const getPublicImg = function (req, res) {
```
dirty method, needs cleanup

mnkami commented on src/server/api/users/usersRoutes.js:
```diff
@@ -90,6 +122,12 @@ router.post('/add', (req, res) => {
 
 // edit user profile
 router.put('/profile/updateUser', (req, res) => {
+    //TODO add comment
```
erm, what comment?

mnkami commented on src/server/api/users/usersRoutes.js:
```diff
@@ -90,6 +122,12 @@ router.post('/add', (req, res) => {
+    //TODO add comment
+    var isUserImgPresent = req.body.userImg && req.body.userImg.image;
+    if (isUserImgPresent) {
+        req.body.userImg.image = req.body.userImg.image.split("/").pop().split('?').shift();
```
wtf?

mnkami commented on src/server/api/users/usersRoutes.js:
```diff
@@ -112,4 +150,49 @@ router.put('/profile/delete', (req, res) => {
+
+    }
+    catch (e) {
+        console.log(e);
```
request will be timed out, you need to handle error

mnkami commented on src/server/chat.js:
```diff
@@ -0,0 +1,37 @@
+
+	(col.find().sort({$natural: -1 }).limit(messageLimit)).toArray(function(err,res){
+	    if(err) 
+	    	throw err;
```
where are you catching it?

mnkami commented on test/client/modules/login/index.spec.js:
```diff
@@ -0,0 +1,105 @@
+
+    beforeEach(angular.mock.module("tanks.login"));
+
+    // describe('Testing sendLog service', function() {
```
commented code

kachanovskyi replied:
got questions on that

mnkami replied:
so ask here, not sure I will be able to attend class tomorrow. Maybe after 2 pm... not sure.

kachanovskyi replied:
think I'll have to wait for You, cause it's a little bit difficult to ask here

mnkami commented on src/client/modules/signup/index.js:
```diff
@@ -1,58 +1,75 @@
+
 import angular from 'angular';
 import ngRoute from 'angular-route';
+var toastr = require('angular-toastr');
```
import!

mnkami commented on src/client/modules/signup/index.js:
```diff
@@ -1,58 +1,75 @@
-}]);
+])
+    .config(RouteConfig)
+    // .config(function (toastrConfig) {
```
any luck with it?

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -397,7 +400,7 @@ const getPublicImg = function (req, res) {
             check();
             fs.readdir(userDir, function (e, files) {
-                console.log(e, files.length > 0);
+                // console.log(e, files.length > 0);
```
just remove it

mnkami commented on test/client/modules/signup/signupSpec.js:
```diff
@@ -0,0 +1,26 @@
+    it("Check for max length of UserName", function () {
+        var $scope = {};
+        var toastr = {};
+        toastr.success = function () {
```
eh?

mnkami replied:
`const toastrMock = {
    success: () => {}
};`

mnkami commented on test/units.html:
```diff
@@ -0,0 +1,194 @@
+<!DOCTYPE html>
```
do you need this file stored in git at all? You will have test task while the build is creating, so it will add all the files you need.

shmoksi committed:
merge with prepullrequest

mnkami commented on src/client/modules/dashboard/index.js:
```diff
@@ -3,17 +3,10 @@ import ngRoute from 'angular-route';
-]).config(RouteConfig);
+])
+.controller('DashboardCtrl', ['$scope', '$http', function($scope, $http) {
+    $scope.var1 = 5;
```
var1? really?

mnkami commented on test/client/modules/dashboard/index.spec.js:
```diff
@@ -0,0 +1,95 @@
+// describe('Test Dashboard', function() {
```
why are you adding commented file to git?

mnkami commented:
Guys, please remove `test/units.html` from git

mnkami commented on src/client/models/externalFunctions.js:
```diff
@@ -0,0 +1,14 @@
+    window.oRequestAnimationFrame      ||
+    window.msRequestAnimationFrame     ||
+    function(callback){
+        window.setTimeout(callback, 1000 / 60);
```
you could omit window

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -1,7 +1,12 @@
-import constans from './constans';
-var originalPoints = constans.originalPoints;
+import externalVariables from './externalVariables';
+let originalPoints = externalVariables.originalPoints,
```
originalPoints is a const

tymochko replied:
I'm going to modify **originalPoints** after each damage on battleground.
Should it be **const**

mnkami replied:
Seems PR does not showing all the code. That's fine, then

tymochko replied:
[generateDamage.js](https://github.com/tymochko/pocket-tanks/blob/prepullrequest/src/client/models/generateDamage.js) was checked by you earlier I guess

mnkami commented on src/client/models/tankMovement.js:
```diff
@@ -25,9 +30,9 @@ module.exports.findLinePoints = function(posX) {
-    timePassed = time - start;
+const animate = (time) => {
+    const duration = 1500;
+    let timePassed = time - start;
```
const

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -1,35 +1,38 @@
+    tankY = externalVariables.tankObj.tankY,
+    angleWeapon = externalVariables.tankObj.angleWeapon;
+
+const WIDTH = externalVariables.WIDTH,
```
wow!
import externalVariables from './externalVariables'; could be replaced by
import { originalPoints, ...  }from './externalVariables';

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -1,35 +1,38 @@
 'use strict';
 import paper from 'paper';
-import constans from './constans';
+import externalVariables from './externalVariables';
```
I'd use abs path in the import.

mnkami commented on src/client/models/tankShot.js:
```diff
@@ -1,35 +1,38 @@
 var ctx;
-var tankX, tankY;
-var angleWeaponInc = 0;
+// var tankX, tankY;
```
commented code

mnkami commented on src/client/models/externalVariables.js:
```diff
@@ -0,0 +1,15 @@
+module.exports.originalPoints = [[0, 280],[200, 350], [350, 150], [500, 250],[800, 250],
+[800, 500],[0, 500],[0, 280]];
+
+module.exports.WIDTH = 800;
```
this style is more server side one. in FE we do just

export const WIDTH = 800;

or 

export default {
 WIDTH: 800
}

kachanovskyi committed:
merged with prepull

mnkami commented on src/client/modules/profile/index.js:
```diff
@@ -1,205 +1,23 @@
-            })
-        }
-    }])
+    .controller('ManageProfileController', ManageProfileController)
```
nice!

kachanovskyi committed:
finished merging with prepull

mnkami commented on src/client/modules/signup/index.js:
```diff
@@ -1,19 +1,13 @@
 import angular from 'angular';
 import ngRoute from 'angular-route';
-var toastr = require('angular-toastr');
+import {toastr} from 'angular-toastr';
```
please teach your teammates how to do this.. it's very handy

mnkami commented on karma.conf.js:
```diff
@@ -109,6 +104,8 @@ module.exports = function (config) {
         concurrency: Infinity,
 
 
+        browserNoActivityTimeout: 300000,  //5 minute for running tests
```
they should execute like in a minute or two :)

mnkami commented on src/server/api/users/usersController.js:
```diff
@@ -276,6 +276,9 @@ const getUserImage = function (req, res) {
         cb(err, foundUser);
     });
     function cb(err, foundUser) {
+        if(foundUser === null) {
```
optional `if(!foundUser) {`

mnkami commented:
This PR is too large! Merge it when the build green.

tymochko committed:
Merge branch 'prepullrequest' into module_send_data

tymochko committed:
Fix backend route, clean up / refactor code

- deleteUser backend route: destroy session;
- clean up code (to ES6 and code style);

kachanovskyi committed:
LVNODQIR-78 finished login tests

kachanovskyi committed:
Merge branch 'prepullrequest' into navigation

OlgaLyubin committed:
merge with prepull

shmoksi committed:
add 3 test for navpanel

shmoksi committed:
merge test_navpanel & prepullrequest

OlgaLyubin committed:
merge with prepull

tymochko committed:
LVNODQIR-92 implement raw template of service to start from

Commit:
Implemented invitation mechanism.

Commit:
Merge branch 'prepullrequest' into LVNODQIR-79-Test-dashboard

OlgaLyubin committed:
merge with prepull

shmoksi committed:
Merge with prepullrequest

OlgaLyubin committed:
Merge branch 'prepullrequest' into tankdrawing

kachanovskyi committed:
LVNODQIR-79 fit -> it

kachanovskyi committed:
package.json: gulp test

tymochko committed:
LVNODQIR-92 set gameCtrl into separate file

tymochko committed:
LVNODQIR-92 add $http get request for testing routes

tymochko committed:
LVNODQIR-90 refactor into module

OlgaLyubin committed:
added trajectory tests

maksym08 committed:
temp changes

maksym08 committed:
LVNODQIR-50 fixed ctx

OlgaLyubin committed:
fixed trajectory tests

maksym08 committed:
LVNODQIR-50 fix initCanvas

Conflicts:
	src/client/models/tankShot.js
	src/client/modules/game/index.js
	src/client/modules/index.js

tymochko committed:
LVNODQIR-50 fix Explosion, shotTrajectory Merge 'origin/fix_canvas'

Conflicts:
	src/client/models/tankShot.js

kachanovskyi committed:
fixed tank movement

kachanovskyi committed:
userRoutes changed let to var

kachanovskyi committed:
merged with second_tank

kachanovskyi committed:
Created tank model

tymochko committed:
Refactor explosion; started to modulate "generateDamage"

OlgaLyubin committed:
changed registration and finished trajectory tests

tymochko committed:
LVNODQIR-69 refactor into separate module and working

kachanovskyi committed:
LVNODQIR-46 modifying tank movement

kachanovskyi committed:
LVNODQIR-46 merged with prepull

kachanovskyi committed:
LVNODQIR-46 fillBackground and clear into externalFunctions

kachanovskyi committed:
LVNODQIR-97

tymochko committed:
Refactor constants export, clean up logs

maksym08 committed:
create sockets on moving

tymochko committed:
Merge branch 'second_tank'

Conflicts:
	src/client/models/explosion.js
	src/client/models/shotTrajectory.js

kachanovskyi committed:
added canvas multi-layerness and created canvas model

tymochko committed:
LVNODQIR-69 fix bug "property of 0"
/ implement logic for findLineSegmentCoefficient for certical
or horizontal line segment occurances

OlgaLyubin committed:
final fix of registration and tests

OlgaLyubin committed:
merge with second_tank

tymochko committed:
LVNODQIR-69 fix bug when last segment of damage was larger then others

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

OlgaLyubin committed:
merge with second_tank

OlgaLyubin committed:
separated drawTank function into a model

tymochko committed:
LVNODQIR-92 separate service into module

Commit:
LVNODQIR-79 common game for 2 players

Commit:
Merge branch 'prepullrequest' into LVNODQIR-79-Test-dashboard

Commit:
merge

tymochko committed:
Implement socket, Merge branch 'module_send_data'

fix and merge

Conflicts:
	src/client/models/shotTrajectory.js
	src/client/models/tankMovement.js
	src/client/models/tankShot.js
	src/client/modules/game/GameCtrl.js
	src/client/modules/game/index.js
	test/client/modules/dashboard/test.spec.js

tymochko committed:
Fix ground disappear after hit target

OlgaLyubin committed:
new architecture for signup module

OlgaLyubin committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

tymochko committed:
LVNODQIR-92 started to work with sockets and service

tymochko committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

shmoksi committed:
all tests

shmoksi committed:
add gulp test to package.json

Commit:
deleted comments

maksym08 committed:
fix tests

shmoksi committed:
Merge

Commit:
Merge branch 'LVNODQIR-79-Test-dashboard' into second_tank

shmoksi committed:
Add rotate tank Angle

shmoksi committed:
Merge branch 'test_navpanel' into second_tank

Commit:
Fixed Dashboard test.

Commit:
Merge branch 'second_tank' into LVNODQIR-79-Test-dashboard

Commit:
Fixed test.

Commit:
Typo.

shmoksi committed:
add rotate weapon

maksym08 committed:
fixed socket

maksym08 committed:
fixed socket

maksym08 committed:
fixed socket

kachanovskyi committed:
added game images to slider

kachanovskyi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

MykhailoDzedzei committed:
new directive for password and deleted some console.log

shmoksi committed:
fix angles tank and weapon on socket

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

# Conflicts:
#	src/server/chat.js

shmoksi committed:
Merge branch 'prepullrequest' of https://github.com/tymochko/pocket-tanks into second_tank

# Conflicts:
#	src/client/models/tankMovement.js
#	test/client/modules/dashboard/test.spec.js
#	test/client/modules/login/index.spec.js
#	test/client/modules/point_analysis/index2.spec.js
#	test/client/modules/tank's_movement/index.spec.js

tymochko committed:
LVNODQIR-92 refactor client invites into dashboard controller

MykhailoDzedzei committed:
Added translation into Ukrainian language

MykhailoDzedzei committed:
LVNODQIR-103 translation of profile page

MykhailoDzedzei committed:
ESLint with new rules

MykhailoDzedzei committed:
fix about translate

tymochko committed:
LVNODQIR-92 implement service into game invites

tymochko committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

tymochko committed:
LVNODQIR-92 refactor service into object

OlgaLyubin committed:
LVNODQIR-101 Created route for scores page

tymochko committed:
LVNODQIR-92 implement game creation with embed users ids

tymochko committed:
LVNODQIR-92 working on simultanious redirect to game

- changes to eslint rules

kachanovskyi committed:
LVNODQIR-99: implemented tank class

kachanovskyi committed:
merged with second_tank

tymochko committed:
LVNODQIR-92 implement receive game data on init stage

kachanovskyi committed:
LVNODQIR-99: fixed tank shot

kachanovskyi committed:
LVNODQIR-99: merged with second_tank

kachanovskyi committed:
added player model

kachanovskyi committed:
Split home and navigation modules into files

tymochko committed:
LVNODQIR-92 fix originalPoints to receive filled in

MykhailoDzedzei committed:
LVNODQIR-106 created instantiate classes

MykhailoDzedzei committed:
resolved conflict inside gameCtrl

kachanovskyi committed:
fixed bullet hittin ground

MykhailoDzedzei committed:
LVNODQIR-106 added socket for transfer and update data into DB

kachanovskyi committed:
working on playersId

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

kachanovskyi committed:
merged with tank_movement

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

MykhailoDzedzei committed:
LVNODQIR-103 dashboard page translated

kachanovskyi committed:
LVNODQIR-99: working on second tank. Implemented different players IDs

tymochko committed:
Extend functionality for ES2015 in NodeJS

tymochko committed:
Merge branch 'second_tank' into service_gameData

tymochko committed:
refactoring socket on shot

tymochko committed:
refactoring socket on shot

MykhailoDzedzei committed:
translation fixes on Navigation

MykhailoDzedzei committed:
merge with second tank

maksym08 committed:
refactoring socket on shot

kachanovskyi committed:
LVNODQIR-99: two tanks on gamefield, working on split tank moves

maksym08 committed:
`Merge branch 'second_tank' into chat2

maksym08 committed:
Merge branch 'second_tank' into chat2

MykhailoDzedzei committed:
LVNODQIR-103 finished translation of the game pages

MykhailoDzedzei committed:
pull second tank

MykhailoDzedzei committed:
LVNODQIR-103 merge with translate_page

MykhailoDzedzei committed:
translation key loading on starting page

kachanovskyi committed:
LVNODQIR-99: tanks move in both windows

kachanovskyi committed:
merged with second_tank

maksym08 committed:
temp changes

maksym08 committed:
refactoring socket

Commit:
modal window for confirm

tymochko committed:
LVNODQIR-107 add button "surrender" and styles

Conflicts:
	src/client/models/shotTrajectory.js

kachanovskyi committed:
LVNODQIR-99: working on two tanks movement on both screens

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

tymochko committed:
LVNODQIR-107 refactor sockets

tymochko committed:
Merge branch 'second_tank' into service_gameData

Conflicts:
	src/client/models/externalFunctions.js

MykhailoDzedzei committed:
bullet intersect second tank

MykhailoDzedzei committed:
LVNODQIR-108 refactoring code and hide bullet after intersection

tymochko committed:
LVNODQIR-107 implement update in DB

tymochko committed:
Merge branch 'second_tank' into service_gameData

tymochko committed:
LVNODQIR-107 fix DB update

tymochko committed:
LVNODQIR-107 implement "SURRENDER" button

- redirects to Dashboard Page to both players
- fix Invitation modal to display user

kachanovskyi committed:
LVNODQIR-99: correct tank movement on both screens (needs refactor)

kachanovskyi committed:
LVNODQIR-99: merged with second tank

kachanovskyi committed:
LVNODQIR-99: fully merged with second_tank

tymochko committed:
LVNODQIR-109 implement player turn (move, shot)

- move chat button call to navPanel
- removed rule from eslint

kachanovskyi committed:
LVNODQIR-99: weapon moves on both screens respectively to the tanks

tymochko committed:
Merge branch 'second_tank' into service_gameData

Conflicts:
	src/client/models/externalFunctions.js
	src/client/models/tankShot.js

MykhailoDzedzei committed:
working on resume game

maksym08 committed:
merge second_tank

maksym08 committed:
merge second_tank

maksym08 committed:
fixed tank shot

MykhailoDzedzei committed:
LVNODQIR-112 resumme game after disconnect

maksym08 committed:
fixed chat

maksym08 committed:
fixed socket.js

kachanovskyi committed:
NavigationConstructorService modified

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

MykhailoDzedzei committed:
LVNODQIR-113 created socket for update data and LVNODQIR-112 resume game works

MykhailoDzedzei committed:
merge with second_tank

tymochko committed:
LVNODQIR-109 implement player highlight in turn

tymochko committed:
Merge branch 'second_tank' into service_gameData

Conflicts:
	src/client/models/tankShot.js
	src/server/socket/gameSocket.js

MykhailoDzedzei committed:
Fixed intersection bullet with tanks

MykhailoDzedzei committed:
merge with second tank

tymochko committed:
LVNODQIR-109 implement turn for weapon

tymochko committed:
LVNODQIR-109 implement change turn after battlefield damage

MykhailoDzedzei committed:
merge with second tank

maksym08 committed:
fixed ingame navpanel

maksym08 committed:
merge second_tank

maksym08 committed:
merge second_tank

MykhailoDzedzei committed:
fix turn

MykhailoDzedzei committed:
merge with second tank

shmoksi committed:
fix reload page SignUp

shmoksi committed:
refactor code

shmoksi committed:
fix session after register

shmoksi committed:
fix redirect after logout

shmoksi committed:
fix style: remove white line before footer & keep footer to bottom

MykhailoDzedzei committed:
fixed socket for update data in DB and painting map synchronously on both screens

shmoksi committed:
add style for lifebar

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into lifebar

shmoksi committed:
add logic for lifebar

tymochko committed:
LVNODQIR-109 refactor changeTurn function

tymochko committed:
LVNODQIR-109 Merge || fix turn highlight

Conflicts:
	src/client/models/externalFunctions.js
	src/client/models/shotTrajectory.js
	src/client/models/tankShot.js

OlgaLyubin committed:
LVNODQIR-101 scores page

OlgaLyubin committed:
LVNODQIR-101 scores page

shmoksi committed:
fix style

shmoksi committed:
Merge branch 'lifebar' into second_tank

tymochko committed:
Refactor shotTrajectory

tymochko committed:
Merge branch 'second_tank' into service_gameData

kachanovskyi committed:
refactored tankShot.js

kachanovskyi committed:
merged with second_tank

maksym08 committed:
fixed chat

maksym08 committed:
merge second_tank

Commit:
chat styles (first part)

kachanovskyi committed:
fixed logination (removed login and password checks)

shmoksi committed:
remove magic number & change color life

shmoksi committed:
Merge branch 'lifebar' into second_tank

# Conflicts:
#	src/client/models/externalVariables.js

Commit:
Merge

shmoksi committed:
maybe change session

Commit:
Merge branch 'second_tank' into LVNODQIR-79-Test-dashboard

Commit:
finished chat styles

shmoksi committed:
add end game, when life ended & socket lifebar

Commit:
modal windows

tymochko committed:
LVNODQIR-109 implement modal on surrender

tymochko committed:
Merge branch 'second_tank' into service_gameData

Conflicts:
	src/client/modules/game/GameCtrl.js
	src/client/modules/game/index.js

OlgaLyubin committed:
LVNODQIR-101 scores

OlgaLyubin committed:
Merge branch 'second_tank' into tankdrawing

kachanovskyi committed:
fixed invitations

kachanovskyi committed:
Merge branch 'second_tank' into tank_movement

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into lifebar

# Conflicts:
#	src/server/socket/gameSocket.js

shmoksi committed:
Revert "maybe change session"

This reverts commit 225fdbb70e68574f709c42dd83eadea6aa1157d9.

OlgaLyubin committed:
LVNODQIR-101 scores page

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into lifebar

OlgaLyubin committed:
Merge branch 'second_tank' into tankdrawing

OlgaLyubin committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into second_tank

Commit:
fixed chat styles

Commit:
Merge branch 'second_tank' into LVNODQIR-79-Test-dashboard

shmoksi committed:
remove ^

shmoksi committed:
rename img

maksym08 committed:
refactoring shotTrajectory, chatField

maksym08 committed:
Merge branch 'second_tank' into chat2

MykhailoDzedzei committed:
resolved some errors from ESLint

MykhailoDzedzei committed:
merge with second tank

Commit:
formatted chat date

Commit:
Merge branch 'second_tank' of github.com:tymochko/pocket-tanks into second_tank

Commit:
Merge branch 'second_tank' into LVNODQIR-79-Test-dashboard

MykhailoDzedzei committed:
refactoring code and translate scores page

MykhailoDzedzei committed:
merge with second tank

MykhailoDzedzei committed:
fixed directive for password

OlgaLyubin committed:
LVNODQIR-101 fixed bugs

OlgaLyubin committed:
LVNODQIR-101 fixed some bugs

shmoksi committed:
step to modal window

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into lifebar

# Conflicts:
#	src/client/models/shotTrajectory.js
#	src/client/modules/game/gameService.js

MykhailoDzedzei committed:
fixed bug with translation in scores page

shmoksi committed:
modalWindow about status game

shmoksi committed:
Merge branch 'second_tank' of https://github.com/tymochko/pocket-tanks into lifebar

Commit:
Merge branch 'second_tank' of github.com:tymochko/pocket-tanks into second_tank

maksym08 committed:
refactoring

maksym08 committed:
refactoring

maksym08 committed:
refactoring

Commit:
Merge branch 'second_tank' of github.com:tymochko/pocket-tanks into second_tank

Commit:
refactoring

MykhailoDzedzei committed:
fixed upload img

MykhailoDzedzei committed:
fixed bug with images

shmoksi committed:
redirect to home page after middleware

shmoksi committed:
Merge branch 'lifebar' into second_tank

MykhailoDzedzei committed:
fix

OlgaLyubin committed:
Update index.spec.js

MykhailoDzedzei committed:
fix

MykhailoDzedzei committed:
merge with second tank

shmoksi committed:
change won and lost

shmoksi committed:
Merge branch 'prepullrequest' of https://github.com/tymochko/pocket-tanks into prepullrequest
